clippy_lints/src/create_dir.rs

```diff
@@ -0,0 +1,50 @@
+use crate::utils::{match_qpath, paths, snippet, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks usage of `std::fs::create_dir` and suggest using `std::fs::create_dir_all` instead.
+    ///
+    /// **Why is this bad?** Sometimes `std::fs::crate_dir` is mistakenly chosen over `std::fs::create_dir_all`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// std::fs::create_dir("foo")
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// std::fs::create_dir_all("foo")
+    /// ```
+    pub CREATE_DIR,
+    restriction,
+    "calling `std::fs::create_dir` instead of `std::fs::create_dir_all`"
+}
+
+declare_lint_pass!(CreateDir => [CREATE_DIR]);
+
+impl LateLintPass<'_> for CreateDir {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Call(ref func, ref args) = expr.kind;
+            if let ExprKind::Path(ref path) = func.kind;
+            if match_qpath(path, &paths::STD_FS_CREATE_DIR);
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    CREATE_DIR,
+                    expr.span,
+                    "calling `std::fs::create_dir` where there may be a better way",
+                    "consider calling `std::fs::create_dir_all` instead",
+                    format!("std::fs::create_dir_all({})", snippet(cx, args[0].span, "..")),
```

Comment:
This line of code is currently problematic as it produces a string starting with `std::fs::` even when users have written `use std::fs::create_dir_all` or something. Are there handy functions or idioms for solving this?

Reply:
i woudl just use `create_dir_all()` here.

Reply:
It turned out that we can't make the tests pass when suggesting `create_dir_all`. I think it's reasonable to suggest `std::fs::create_dir_all` and beg for users to adjust it by hand if they've written `use ...` in their code.

Reply:
I mean, we can just make the tests use `std::fs` and assume that's the case in the suggestion. That's why it's MaybeIncorrect. The primary goal of the suggestion is to be concise and helpful, if it's not automatically applicable, it's not, no biggie.

Reply:
Now I get the picture, thanks for your clarification.